### ISSUE TITLE
[Feature request] Support for causal mask flash attention when seq_len_q != seq_len_k

### ISSUE THREAD

**Issue by beginlner** [Dao-AILab/flash-attention#402]
During inference when seq_len_q != seq_len_k, the causal mask should be like this
https://github.com/facebookresearch/llama/blob/main/llama/model.py#L279-L282
Based on my tests, the current flash attention does not work this way.

**Comment by kklemon** [comment]
I think this is not natively supported at the moment. As a workaround, you can perform causal self-attention between the concatenation of the memory and input sequence and slice out the memory. This should have the same effect but will most likely be less efficient than a native implementation of this feature.

```python
import torch
from flash_attn import flash_attn_qkvpacked_func

def causal_attn_with_memory(qkv, qkv_memory):
    mem_len = qkv_memory.size(1)

    input = torch.cat([qkv_memory, qkv], dim=1)

    # Each element from qkv will attend to all elements from qkv_memory
    # but not to _future_ tokens within qkv.
    out = flash_attn_qkvpacked_func(input, causal=True)

    # Output has shape (batch_size, mem_len + seq_len, ...)
    # We are only interested in the attention result for the query tokens.
    return out[:, mem_len:]

torch.set_default_device('cuda')
torch.set_default_dtype(torch.float16)

batch_size = 32
seq_len = 128
mem_len = 64

n_heads = 4
head_dim = 128

qkv = torch.randn(32, seq_len, 3, n_heads, head_dim)
qkv_mem = torch.randn(32, mem_len, 3, n_heads, head_dim)

out = causal_attn_with_memory(qkv, qkv_mem)

assert out.shape == (batch_size, seq_len, n_heads, head_dim)
```

_Memory_ in this case refers to the sequence that the queries should additionally attend to and should not be masked out. This could, for instance, be a prompt or some additional context.

**Comment by low-flyer** [comment]
This is indeed a method. but in a typical situation, such as when mem_len=1000 and seq_len=100, the natively supported flash attention would be much more efficient than this method.

**Comment by tridao** [comment]
We'll support this kind of causal mask eventually (hopefully soon, but I can't promise), so that it's efficient for this use case.

**Comment by beginlner** [comment]
Thank you. This will be very helpful.

**Comment by leizhao1234** [comment]
```python
query = torch.rand((batch_size, seq_len, num_head, hidden_units), dtype=torch.float16, device="cuda")
key = torch.rand((batch_size, seq_len+1, num_head, hidden_units), dtype=torch.float16, device="cuda")
value =  torch.rand((batch_size, seq_len+1, num_head, hidden_units), dtype=torch.float16, device="cuda")
```
However, through my testing, I found that flash-attention supports a causal mask where seq_q is not equal to seq_k.

**Comment by tridao** [comment]
This is now implemented.
